examples/llm/src/mosaic_gpt.py

```diff
@@ -488,10 +489,26 @@ def loss(self, outputs, batch):
     def get_metrics(self, is_train=False):
         return self.train_metrics if is_train else self.eval_metrics
 
-    def update_metric(self, batch, outputs, metric):
-        outputs = outputs.view(-1, outputs.size(-1))
-        targets = self.get_targets(batch).view(-1)
-        metric.update(outputs, targets)
+    def update_metric(self, batch, outputs, metric) -> None:
+        if isinstance(metric, InContextLearningMetric):
+            if batch.get('mode', None) == 'icl_task':
+                # only apply ICL metrics to specially constructed
+                # icl_task batches
+                targets = self.get_targets(batch)
+                metric.update(batch, outputs, targets)
+        else:
+            outputs = outputs.view(-1, outputs.size(-1))
+            targets = self.get_targets(batch).view(-1)
+            metric.update(outputs, targets)
```

Comment:
This code block suggests that if the batch has `mode == 'icl_task'`, it can still be used to update other metrics like `Perplexity`... is that right?

Reply:
I check the type of the metric as well on line 493, so it couldn't trigger for perplexity!

Reply:
Sorry what I mean is, if the `metric` is Perplexity and the `batch.mode == 'icl_task'`, won't it get to line 500? And is that intentional?

Reply:
I think that nothing prevents someone from messing up in that way, but the expectation is that you won't define an `Evaluator` that uses `icl_task` mode, and also attach a non `InContextLearningMetric` to that evaluator. We could raise a nicer error than whatever happens now if you try to do that though.